Destroying the greentext item properly restores players' colors

FlufflesTheDog opened the pull request:
## About The Pull Request
Title. Colors appear to always be stored in lowercase, so the check against uppercase text was missing it. Also changed the added colors to lowercase for consistency. 
Though comparing colors like this in the first place feels a bit hacky, so I'm open to other suggestions
## Why It's Good For The Game
Fix bugs!
## Changelog
:cl:
fix: Destroying the fabled "green text" now properly releases its victims from their curse.
/:cl:

FlufflesTheDog committed:
Greentext destruction fix

Now properly releases those afflicted by its curse when destroyed

FlufflesTheDog committed:
Fix silly variable name

Might as well while I'm here. This variable isn't every single player,
it's only one player.

san7890 commented on code/modules/events/wizard/greentext.dm:
```diff
@@ -75,21 +75,21 @@
-		if(all_player_mobs in color_altered_mobs)
+		if(player in color_altered_mobs)
 			message += " and you're finally able to forgive yourself"
-			if(all_player_mobs.color == "#FF0000" || all_player_mobs.color == "#00FF00")
-				all_player_mobs.remove_atom_colour(ADMIN_COLOUR_PRIORITY)
+			if(player.color == "#ff0000" || player.color == "#00ff00")
+				player.remove_atom_colour(ADMIN_COLOUR_PRIORITY)
 		message += "...</span>"
 		if(!quiet)
-			to_chat(all_player_mobs, message)
+			to_chat(player, message)
```
we can make the whole string concatenation procedure a lot less convoluted, just like this.

```dm
		var/list/messages = list(span_warning("A dark temptation has passed from this world!"))
		if(player in color_altered_mobs)
			messages += span_green("You're finally able to forgive yourself...")
			if(player.color == "#ff0000" || player.color == "#00ff00")
				player.remove_atom_colour(ADMIN_COLOUR_PRIORITY)
		if(!quiet)
			to_chat(player, messages.Join(" "))
```

Reworking this as a list of strings that gets `Join()`'d at the very end is (allegedly) a bit faster than doing base string concatenation, and it allows you that greater flexibility over the spans in the strings (which make it look a bit cooler imo). We can also use the span macros, much cleaner. This is optional since you didn't really touch this code but I wanted to mention this anyways.

FlufflesTheDog replied:
I used "\n" on the join because imo it looks a bit weird to have a color change mid line

san7890 commented on code/modules/events/wizard/greentext.dm:
```diff
@@ -75,21 +75,21 @@
+		last_holder.add_atom_colour("#ff0000", ADMIN_COLOUR_PRIORITY)
 		last_holder = new_holder //long live the king
 
 /obj/item/greentext/Destroy(force)
 	LAZYREMOVE(SSticker.round_end_events, roundend_callback)
 	QDEL_NULL(roundend_callback) //This ought to free the callback datum, and prevent us from harddeling
```
oh btw this shouldn't be in `Destroy()`, because we want that to go by ASAP and a `for()` loop that handles cleanup not related to reference handling is quite poor practice. Would you mind moving everything from lines 84-92 into a new proc on `/obj/item/greentext` and then doing an `INVOKE_ASYNC()` so we don't clog up the `Destroy()` chain doing work that should be done elsewhere? If it's not critical to cleaning up the object before we delete it, it's best to ASYNC it out. I know it's not something you really touched here but I think it's pretty important to be done since this can really make the server weep for a little bit.

san7890 reviewed (changes requested):
nits that are mostly optional but should still be done by _someone_.

san7890 commented:
> Though comparing colors like this in the first place feels a bit hacky, so I'm open to other suggestions

You can make a new trait and then use `ADD_TRAIT`/`HAS_TRAIT` and that would work better.

FlufflesTheDog commented:
> You can make a new trait and then use `ADD_TRAIT`/`HAS_TRAIT` and that would work better. in fact it would work best in case other ways to edit player's colors interfere with this stuff, whereas the trait would be completely pure.

The actual color system has a priority list, so I believe only something with the same priority would have any chance of being trampled either way. But would a trait actually help with that? It would tell you for certain that someone was recolored by the item, but not if they've been recolored by something else in the mean time, for example.

FlufflesTheDog commented:
The more I think about this the more I'm not sure what the "correct" way to do it would be.

I realized two other potential problems, being that the list of mobs opens up lots of chances for harddels, and even though we're storing a list of mobs that have been recolored, we only use it for verification, so dead and clientless mobs don't get recolored.

It **sounds** reasonable to iterate over the tracked list instead, but I'm not sure how to fix the first point.

san7890 commented:
do this

```dm
/datum/thing
     /// lazylist of stuff that we need to use later
     var/list/stuff = null

/datum/thing/proc/perform(mob/target)
     LAZY_ADD(stuff, target) // or whatever the syntax is to add target to stuff as a lazylist

/datum/thing/Destroy()
    INVOKE_ASYNC(cleanup_the_list, stuff.Copy()
    stuff = null
    return ..()

/// Handles all of the logic for cleaning up our list
/datum/thing/proc/cleanup_the_list(list/list_of_stuff)
    for(var/x in list_of_stuff)
          // do whatever here.
```

you could also use a list of weakrefs instead but i think then we're going to silly town.

FlufflesTheDog commented:
Okay, that helps. Though a couple more questions:

Is it required to copy the list when passing it to cleanup? I was under the impression that gc would handle that okay, and only delete the list once both things are done with it.

And does it matter if src is null (or qdelled, hypothetically?) Since you don't pass src in invoke_async, and this function is being called after Destroy() even if you did

FlufflesTheDog committed:
Scooch this into async land

FlufflesTheDog committed:
Span macros

FlufflesTheDog committed:
Announce and recolor seperately

Removes the color check since it's a bit redundant
Recolors dead & clientless mobs
Silent version will still announce to effected players

FlufflesTheDog committed:
Use traits for checking if someone is afflicted

FlufflesTheDog committed:
Weakrefs! Stops recolored mobs from harddelling

FlufflesTheDog committed:
Scraps a bunch of redundant checks

It SHOULDN'T be possible for someone to equip this without the last
person dropping it. Instead of wasting processing time being paranoid,
lets treat exceptions as bugs and fix them.

FlufflesTheDog committed:
Check for objectives AFTER checking for mind

Fixes a runtime on mobs with no mind

FlufflesTheDog committed:
One more span macro

FlufflesTheDog committed:
whoopsie

san7890 commented:
> Is it required to copy the list when passing it to cleanup? I was under the impression that gc would handle that okay, and only delete the list once both things are done with it.

yes because we're invoking it ASYNC. the list will have been deleted and we won't be able to do any work with the list since the very next line will null it out, meaning we get an effectively null reference passed to the clean up proc

> And does it matter if src is null (or qdelled, hypothetically?) Since you don't pass src in invoke_async, and this function is being called after Destroy() even if you did

make the cleanup proc a global then just `/proc/cleanup_list_of_stuf()`, no need to rely on the item itself after that point, just gotta get those things cleaned up.

FlufflesTheDog committed:
Release victims before congratulating the winner

Same result, but makes the "you've won" stuff show up after the "evil
has passed from this world" flavor, which feels a bit more fitting, and
puts the more relevant information for the winner lower in the chat.

FlufflesTheDog committed:
appease the lint

FlufflesTheDog commented:
I already had it written out without the global proc or list copy. It does seem to work fine, haven't been getting any runtimes or hard dels, but I could add those changes in if you're sure this wouldn't be reliable.

san7890 commented:
> I already had it written out without the global proc or list copy. It does seem to work fine, haven't been getting any runtimes or hard dels, but I could add those changes in if you're sure this wouldn't be reliable.

i'm concerned about it potentially hanging in the future but i think now that it's a list of weakrefs instead of a list of mobs we should be okay now:tm:

thanks for sticking with the process and doing a really big overhaul of the code instead of the bare fix you set out to do, really means a lot.

FlufflesTheDog commented:
It was a good learning experience, thanks for the help getting it done.
